lib/web_ui/lib/src/engine/platform_dispatcher/app_lifecycle_state.dart

```diff
@@ -0,0 +1,106 @@
+typedef AppLifecycleStateListener = void Function(ui.AppLifecycleState state);
+
+/// Determines the [ui.AppLifecycleState].
+abstract class AppLifecycleState {
+  static final AppLifecycleState instance = _BrowserAppLifecycleState();
+
+  ui.AppLifecycleState get appLifecycleState => _appLifecycleState;
+  ui.AppLifecycleState _appLifecycleState = ui.AppLifecycleState.resumed;
+
+  final List<AppLifecycleStateListener> _listeners =
+      <AppLifecycleStateListener>[];
+
+  void addListener(AppLifecycleStateListener listener) {
+    if (_listeners.isEmpty) {
+      activate();
+    }
+    _listeners.add(listener);
+    listener(_appLifecycleState);
+  }
+
+  void removeListener(AppLifecycleStateListener listener) {
+    _listeners.remove(listener);
+    if (_listeners.isEmpty) {
+      deactivate();
+    }
+  }
+
+  @protected
+  void activate();
+
+  @protected
+  void deactivate();
+
+  @visibleForTesting
+  void onAppLifecycleStateChange(ui.AppLifecycleState newState) {
+    if (newState != _appLifecycleState) {
+      _appLifecycleState = newState;
+      for (final AppLifecycleStateListener listener in _listeners) {
+        listener(newState);
+      }
+    }
+  }
+}
+
+/// Manages [ui.AppLifecycleState] within a web context by monitoring specific
+/// browser events.
+///
+/// This class listens to:
+/// - 'beforeunload' on [DomWindow] to detect detachment,
+/// - 'visibilitychange' on [DomHTMLDocument] to observe visibility changes,
+/// - 'focus' and 'blur' on [DomWindow] to track application focus shifts.
+class _BrowserAppLifecycleState extends AppLifecycleState {
+  @override
+  void activate() {
+    domWindow.addEventListener('focus', _focusListener);
+    domWindow.addEventListener('blur', _blurListener);
+    // TODO(web): Register 'beforeunload' only if lifecycle listeners exist, to improve efficiency: https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#usage_notes
+    domWindow.addEventListener('beforeunload', _beforeUnloadListener);
+    domDocument.addEventListener('visibilitychange', _visibilityChangeListener);
+  }
+
+  @override
+  void deactivate() {
+    domWindow.removeEventListener('focus', _focusListener);
+    domWindow.removeEventListener('blur', _blurListener);
+    domWindow.removeEventListener('beforeunload', _beforeUnloadListener);
+    domDocument.removeEventListener(
+      'visibilitychange',
+      _visibilityChangeListener,
+    );
+  }
+
+  late final DomEventListener _focusListener =
+      createDomEventListener((DomEvent event) {
+    onAppLifecycleStateChange(ui.AppLifecycleState.resumed);
+  });
+
+  late final DomEventListener _blurListener =
+      createDomEventListener((DomEvent event) {
+    onAppLifecycleStateChange(ui.AppLifecycleState.inactive);
+  });
+
+  late final DomEventListener _beforeUnloadListener =
+      createDomEventListener((DomEvent event) {
+    onAppLifecycleStateChange(ui.AppLifecycleState.detached);
```

Comment:
Since this uses detached on all web platforms you'll need to change the documentation for the AppLifecycleState class, both in the engine and for the framework.

https://github.com/gspencergoog/engine/blob/main/shell/platform/common/app_lifecycle_state.h#L36-L37

https://github.com/gspencergoog/engine/blob/main/lib/ui/platform_dispatcher.dart#L1794-L1795